src/adaptations/device-layer/FreeRTOS/SystemTimeSupport.cpp

```diff
@@ -40,24 +40,59 @@ constexpr uint32_t kTicksOverflowShift = (configUSE_16_BIT_TICKS) ? 16 : 32;
 
 uint64_t sBootTimeUS = 0;
 
-uint64_t TicksSinceBoot(void)
+#ifdef __CORTEX_M
+BaseType_t sNumOfOverflows;
+#endif
+} // unnamed namespace
+
+/**
+ * Returns the number of FreeRTOS ticks since the system booted.
+ *
+ * NOTE: The default implementation of this function uses FreeRTOS's
+ * vTaskSetTimeOutState() function to get the total number of ticks,
+ * irrespective of tick counter overflows.  Unfortunately, this function cannot
+ * be called in interrupt context, no equivalent ISR function exists, and
+ * FreeRTOS provides no portable way of determining whether a function is being
+ * called in an interrupt context.  Adaptations that need to use the Weave
+ * Get/SetClock methods from within an interrupt handler must override this
+ * function with a suitable alternative that works on the target platform.  The
+ * provided version is safe to call on ARM Cortex platforms with CMSIS
+ * libraries.
+ */
+
+uint64_t TicksSinceBoot(void) __attribute__((weak))
 {
     TimeOut_t timeOut;
+
+#ifdef __CORTEX_M
+    if (SCB->ICSR & SCB_ICSR_VECTACTIVE_Msk) // running in an interrupt context
+    {
+        // Note that sNumOverflows may be quite stale, and under those
+        // circumstances, the function may violate monotonicity guarantees
+        return static_cast<uint64_t>(xTaskGetTickCountFromISR()) +
+                static_cast<uint64_t>(sNumOfOverflows << kTicksOverflowShift);
```

Comment:
Not a big deal, but you could collapse this with the return statement below by simply setting the fields within timeOut here.

Reply:
True.  I like this idea, it will clean it up